rfcs/0012-mtls.md

```diff
@@ -0,0 +1,106 @@
+# RFC Template
+
+- Feature Name: `mtls`
+- Start Date: 2024-11-27
+- RFC PR: [Kuadrant/architecture#110](https://github.com/Kuadrant/architecture/pull/110)
+- Issue tracking: [Kuadrant/kuadrant-operator#1049](https://github.com/Kuadrant/kuadrant-operator/issues/1049)
+
+# Summary
+[summary]: #summary
+
+MTLS support for communication between the Wasm plugin in the gateway pod, and Authorino & Limitador.
+
+# Motivation
+[motivation]: #motivation
+
+For security reasons so that all traffic is encrypted, and certificates validated.
+This aligns with a 'zero trust' architecture.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Update the Kuadrant CR to have a new configuration option `mtls` for users to configure if mtls should be enabled for either authorino or limitador.
+The reason for having separate flags for either component is to allow flexibility where the resource overhead of tls may be considered too high when enabled for a specific component e.g. for checking limits on each request in limitador.
+The new config in the Kuadrant CR would look like this:
+
+```yaml
+apiVersion: Kuadrant.io/v1alpha1
+kind: kuadrant
+spec:
+  mtls:
+    authorino: true
+    limitador: true
```

Comment:
I think there is a design issue here. More so when we think of the whole CR and just this feature. As this feature only has two components that is affected this will be harder to visualize, but I will try explain it from a generic features point of view.

```yaml
spec:
  feature:
    activate: false # optional, boolean, defaults to false
    enabled: [] # optional, list of strings
    disabled: [] # optional, list of strings
```
Under the feature all fields are optional to ensure backwards compatibility. The feature will not be activated on less the `activate` field is set to true. This allows the feature to be enabled or disabled without changing the configuration for that feature. A feature would have a list of defaults components that would be affected by the feature being enabled. In this case that would be two, `authorino` & `limitador`, but this would not always be the case. 

If the user wanted to disable the feature on a component lets say limitador they would add the following spec.
```yaml
spec:
  feature:
    activate: true 
    disabled: ['limitador']
```
In this example only limitador will not have the feature enabled, but also if any new components are added to the defaults list that will be auto enabled. Being auto enabled by default may not be what the user wants. That is where the `enabled` list comes in. If the user wants to only have the feature enabled for authorino no matter what components are add in the future they could use the following spec.
```yaml
spec:
  feature:
    activate: true 
    enable: ['authorino']
```
The enable list would be an explicit list. Only components listed would be affected. This ensures that behaviour would not change if a component is added to the support list later in the future. However, if there are a lot of components that is supported by default filling this list in could be tedious, but as it is a string list there is nothing stopping use from supporting keywords like `'defaults'`. An example of this being used could be service monitors whereby default we only add them to the operands. A user many also add them to the operator that they care about. The below example shows that looks like.
```yaml
spec:
  feature:
    activate: true
    enable: 
      - 'defaults'
      - 'authorino-operator'
```

So advantage that I see with doing this design are.
- All features have the same basic interface.
- New components don't require CRD updates to be supported. The change will only be required with the operator.
- The common interface make documentation simpler

Reply:
I like this approach.
Are the enabled & disabled lists mutually exclusive?